README.md
What is this?
=============

Record is a photo camera, video camera, and audio recorder for the Sugar desktop.

How to use?
===========

Record is part of the Sugar desktop and is often included.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [Download Record using Browse](https://activities.sugarlabs.org/), search for `Record`, then download, and;
* [How to use Record](https://help.sugarlabs.org/en/record.html).

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/en/my_settings.html), [Software Update](https://help.sugarlabs.org/en/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `Record`, then download.

How to integrate?
=================

Record depends on Python, [Sugar Toolkit for GTK+ 3](https://github.com/sugarlabs/sugar-toolkit-gtk3), and PyGObject bindings for GStreamer 1 and GTK+ 3.

Record is started by [Sugar](https://github.com/sugarlabs/sugar).

Record is [packaged by Fedora](https://src.fedoraproject.org/rpms/sugar-record).  On Fedora systems;

```
dnf install sugar-record
```

Record is no longer packaged by Debian and Ubuntu distributions.  When it was packaged, it was called `sugar-record-activity`.  On Debian and Ubuntu systems dependencies include;

* `python`
* `python-gobject`
* `python-cairo`
* `python-sugar3`
* `gir1.2-glib-2.0`
* `gir1.2-gdkpixbuf-2.0`
* `gir1.2-gtk-3.0`
* `gir1.2-pango-1.0`
* `gir1.2-telepathyglib-0.12`
* `gir1.2-gstreamer-1.0`
* `python-gst-1.0`
* `gstreamer1.0-plugins-base`
* `gstreamer1.0-plugins-good`

How to test without camera
==========================

For virtual machines, set the RECORD_TEST environment variable before
starting the activity, and the GStreamer videotestsrc will be used.

Branch master
=============

The `master` branch targets an environment with latest stable release
of [Sugar](https://github.com/sugarlabs/sugar), with dependencies on
latest stable release of Fedora and Debian distributions.

Branch not-gstreamer1
=====================

The `not-gstreamer1` branch is a backport of features and bug fixes
from the `master` branch for ongoing maintenance of the activity on
Fedora 18 systems which don't have well-functioning GStreamer 1
packages.


NEWS
201

* Port to Python 3 (Srevin Saju),

200.2

* Fix mistaken reference to logging,
* Fix README reference to Speak,
* Adjust license metadata,
* Add tags metadata,
* Port from GObject to GLib,
* tested with GStreamer 1.14 on Ubuntu 18.04,

200.1

* Update Debian and Ubuntu dependencies,
* Fix fullscreen toggle of live view on Ubuntu 16.04,
* Fix fail to start on Ubuntu 16.04.

200

* port to GTK+ 3 and GStreamer 1.0,
* tested with GStreamer 1.14 on Ubuntu 18.04,

103

* fix playback seeking on progress bar,
* avoid double JPEG compression of photographs,
* add tooltips to mode buttons, shutter button, and pause and play,
* new shortcut keys; ctrl+space to record, escape to live mode, ctrl+1
  through to ctrl+3 for modes, ctrl+r for record, space for pause and
  play,
* use an accurate countdown timer,
* switch to external camera if available,
* mirror mode for hairstyling, shortcut key ctrl-m,
* show all recorded media in tray regardless of type.

102

--

* eliminate combo boxes for more friendly to touch UI

76

* Switch to Daniel Drake's branch
  Restore the v60 rewrite to use a single pipeline, tweaking at runtime
  when necessary. This improves responsivity of the application, and results
  in the activity behaving reliably again.
* This version requires gstreamer-0.10.20

68..75

* Range was preserved for possible sucrose-0.82 branch updates

67

* Fix 0sugar "Cannot find cached 0sugar-launch implementation." isuue

66

* Increase audio volume on XO
* 0sugarize Record blobs to let use native gst on recent systems

65

* Use hippo.cairo_surface_from_gdk_pixbuf instead of bundling blobs
* Add gst logging
* Record 64 does not record sound #1244

64

* #848 Hide photo/video tabs if camera doesn't exist 
* Update .po files

63

* Change only volumes for "Capture" channels and let user change others
* Record controls only appear on second launch #762

62

* Select quality for enconding video
* Unmute/set-to-max capture channels
  workaround code until sugar controls capture volumes in the shell
* Revert two pass video encoding to support limited hardware
* Save GUI state in Journal objects

61

* Use uuid and do not md5 video files
* Fallback to ximagesink if xvimagesink is unaccessible
* Use one pass for encodings

60

* pipeline rework; makes the activity work with new gstreamer

59

* #8580 record fails to keep images if closed. (erikg)

58

* Add MANIFEST (cscott)
* Update license and update_url fields (cscott)

54
* #6679 fixed

53
* #6417 fixed

52
* updated translations (jedierikb, sayamindu)

51
* #4525 defensive techniques against translations with bad variables (jedierikb)
* #5899 fixed (jedierikb)
* tag support (jedierikb)

50
* #5830 fix (jedierikb)
* #4525 updates with fixes to .po files (jedierikb)
* #5899 workaround to avoid occasional crashes (jedierikb)

49
* #4525 updated with .po files (jedierikb, sayamindu)

48
* #5500 fix (jedierikb)
* #4525 applied (jedierikb)

47
* #5448 updated (jedierikb)

46
* #5448 fix (jedierikb)
* #4983 addressed (jedierikb)

45
* new .pot file (jedierikb)
* fullscreen flag re-added (jedierikb)

44
* clipboard fix (jedierikb)

43
* #5230 fixed (jedierikb)

42
* #4983 addressed (jedierikb)

41
* #5177, fullscreen compliance for u1 (jedierikb)
* #4813, clipboard files via /instance not /tmp (jedierikb)

40
* filepath changes for bitfrost, rainbow (jedierikb)

39
* tray improvements (jedierikb)
* thumbnail improvements (jedierikb)
* game key improvements (jedierikb)

38
* change time of creation of tmp (jedierikb)

37
* prevent duplicate thumbnails when shared (jedierikb)
* prevent changing modes when using timer (jedierikb, mikhak)
* fix audio mesh share (jedierikb, mikhak)

36
* ui (jedierikb, mikhak, eben)
* tubes

35
* #4422 addressed (jedierikb)

34
* frs ui (jedierikb, mikhak)
* notify::active #2026 should close #3097, #3616 (jedierikb)
* #4111, #4189 fixed (jedierikb)
* #1063, #3776 addressed (jedierikb), blocked by #4144
* #809 fixed (jedierikb)
* #3774 fixed (jedierikb)
* #4209 worked around (jedierikb)
* #1557 addressed (jedierikb)
* #3588 fixed (jedierikb)

33
* bug with slow creation of audio thumbnails (jedierikb)

32
* changed audio/wav mimetype to audio/x-wav for entry into the journal

31
* #4010 audio/wav mimetype for entry into the journal

30
* ogg not ogv

27
* delete (jedierikb)

26
* ui (jedierikb)

25
* fix for narrowly possible gtk bug when taking pictures (closes #3342)
* fix for xv "glitches" when moving and resizing gtk.Windows (closes #3046)
* enable the disappearing record buttons which don't disappear if your moused over them (jedierikb)

24
* fix for saving audio files' image counterparts (prevents blank screen and spinning cursor) (jedierikb)
* correctly handle bug in notify::active callbacks, and opened #2026 (jedierikb)
* disable the disappearing record buttons until more work can be done on them (jedierikb)
* correct mime and filetypes for audio recordings (addresses #2909) (jedierikb)
* fix for hitherto unknown bug regarding multiple saves to the datastore (jedierikb)
* new icon (simon@schampijer.de)

23
* journal integration (closes #1758) (jedierikb):
** media saved to the the journal
** re-naming of media in the record actvity and the journal
** deleting of media in the record activity and the journal
* correct ogg/theora mimetypes (addresses #2653) (jedierikb)
* sharing of all recorded media on the mesh (closes #2436) (jedierikb)
* you can delete buddy's photos pushed to you (closes #2722) (jedierikb)
* you can view buddy's photos when you re-open an activity (closes #2439) (jedierikb)
* all widgets insensitive when shuttingdown and writing to journal (jedierikb)
* better pipeline-to-window hot swapping (jedierikb)
* audio record and playback (jedierikb)
* drag and drop of all media types for all media (closes #2345)(jedierikb, tomeu)
* graphic overlay windows fade away when not used to reveal photo, revealed with mouse movement (jedierikb)
bugs:
* graphic flashes on the resizing of video windows (#3046)

glive.py
# Copyright (C) 2008, Media Modifications Ltd.
# Copyright (C) 2011, One Laptop per Child

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import os
import glob
from gettext import gettext as _
import time

from gi.repository import GLib, Gst, GstVideo

import logging

from instance import Instance
import constants
import utils

logger = logging.getLogger('glive')


class Glive:
    def __init__(self, activity_obj, model):
        logger.debug('__init__')
        self.activity = activity_obj
        self.model = model

        self._cameras = sorted(glob.glob('/dev/v4l/by-id/*index0'))
        if os.getenv('RECORD_TEST'):
            self._cameras.append('test')

        if self._cameras:
            self._camera = self._cameras[0]

        self._pixbuf = None
        self._audio_pixbuf = None
        self._video_pixbuf = None

        self._xid = None
        self._sink = None

        self._mirror = False

        self._pipeline = self._make_photo_pipeline()

    def get_cameras(self):
        return self._cameras

    def switch_camera(self):
        which = ((self._cameras.index(self._camera)) + 1) % len(self._cameras)
        self._camera = self._cameras[which]
        self.stop()
        self._pipeline = self._make_photo_pipeline()
        self.play()

    def set_mirror(self, mirror):
        if self._mirror != mirror:
            self._mirror = mirror
            self._set_video_direction(self._pipeline)

    def _set_video_direction(self, pipeline):
        flip = pipeline.get_by_name('flip')
        value = 4 if self._mirror else 0
        if hasattr(flip.props, 'method'):
            flip.props.method = value
        if hasattr(flip.props, 'video_direction'):
            flip.props.video_direction = value

    def _make_photo_pipeline(self):
        """
        create a Gst.Pipeline for
        - displaying camera video on display, with optional mirroring,
        - capturing photographs,
        """

        if self._cameras:
            args = {'src': 'v4l2src device={0}'.format(self._camera)}
            if self._camera == 'test':
                args = {'src': 'videotestsrc'}
            args['cap'] = ''
        else:
            args = {'src': 'videotestsrc pattern=black',
                    'cap': '! video/x-raw,framerate=5/1,width=640,height=480 '}

        cmd = '{src} name=src {cap}' \
            '! videorate max-rate=10 ' \
            '! tee name=tee ' \
            'tee.! videoconvert ! queue leaky=2 ' \
            '! videoflip name=flip ' \
            '! autovideosink sync=false ' \
            'tee.! videoconvert ! queue ! gdkpixbufsink name=photo'

        pipeline = Gst.parse_launch(cmd.format(**args))
        self._set_video_direction(pipeline)

        photo = pipeline.get_by_name('photo')

        # catch photos
        bus = pipeline.get_bus()
        bus.add_signal_watch()

        def on_message_cb(bus, msg):
            if msg.get_structure() is not None:
                if msg.get_structure().get_name() == 'pixbuf':
                    self._pixbuf = photo.get_property('last-pixbuf')
                    return

            return self._on_message_cb(bus, msg)

        bus.connect('message', on_message_cb)

        self._catch_window(bus, self.activity.set_glive_sink)
        return pipeline

    def _catch_window(self, bus, cb):
        bus.enable_sync_message_emission()

        def on_sync_message_cb(bus, msg):
            if msg.get_structure().get_name() == 'prepare-window-handle':
                cb(msg.src)

        bus.connect('sync-message::element', on_sync_message_cb)

    def _on_message_cb(self, bus, msg):
        if msg.type == Gst.MessageType.ERROR:
            err, debug = msg.parse_error()
            logger.error('bus error=%s debug=%s' % (err, debug))
            return

        if msg.type == Gst.MessageType.WARNING:
            err, debug = msg.parse_warning()
            logger.error('bus warning=%s debug=%s' % (err, debug))

    def play(self):
        logger.debug('play')
        if self._get_state() == Gst.State.PLAYING:
            return

        self._pipeline.set_state(Gst.State.PLAYING)  # asynchronous

    def stop(self):
        logger.debug('stop')
        if self._get_state() == Gst.State.NULL:
            return

        self._pipeline.set_state(Gst.State.NULL)  # synchronous
        self._pixbuf = None
        self.activity.set_glive_sink(None)

    def _get_state(self):
        return self._pipeline.get_state(Gst.CLOCK_TIME_NONE)[1]

    def take_photo(self):
        logger.debug('take_photo')
        if self._cameras:
            self.model.save_photo(self._pixbuf.copy())

    def record_audio(self):
        logger.debug('record_audio')

        # take a photograph
        self._audio_pixbuf = self._pixbuf.copy()
        self.model.still_ready(self._audio_pixbuf)

        # stop the live view
        self._pipeline.set_state(Gst.State.NULL)  # synchronous
        self._pixbuf = None

        # make a pipeline to record and encode audio to file
        ogg = os.path.join(Instance.instancePath, "output.ogg")
        cmd = 'autoaudiosrc name=src ' \
            '! audioconvert ' \
            '! queue max-size-time=30000000000 ' \
            'max-size-bytes=0 max-size-buffers=0 ' \
            '! vorbisenc name=vorbis ! oggmux ' \
            '! filesink location=%s' % ogg
        self._audio = Gst.parse_launch(cmd)

        # attach useful tags
        taglist = self._get_tags(constants.TYPE_AUDIO)
        if self._audio_pixbuf:
            pixbuf_b64 = utils.getStringEncodedFromPixbuf(self._audio_pixbuf)
            taglist.add_value(Gst.TagMergeMode.APPEND,
                              Gst.TAG_EXTENDED_COMMENT,
                              "coverart=" + str(pixbuf_b64))

        vorbis = self._audio.get_by_name('vorbis')
        vorbis.merge_tags(taglist, Gst.TagMergeMode.REPLACE_ALL)

        # detect end of stream
        bus = self._audio.get_bus()
        bus.add_signal_watch()

        def on_message_cb(bus, msg, ogg):
            if msg.type == Gst.MessageType.EOS:
                logger.debug('record_audio.on_message_cb Gst.MessageType.EOS')
                GLib.idle_add(self._stop_recording_audio, ogg)
                return

            return self._on_message_cb(bus, msg)

        bus.connect('message', on_message_cb, ogg)

        # start audio pipeline recording
        self._audio.set_state(Gst.State.PLAYING)  # asynchronous

    def stop_recording_audio(self):
        logger.debug('stop_recording_audio')

        # ask for stream to end
        self._audio.get_by_name('src').send_event(Gst.Event.new_eos())

    def _stop_recording_audio(self, ogg):
        logger.debug('_stop_recording_audio')

        # save audio file
        self.model.save_audio(ogg, self._audio_pixbuf)
        self._audio_pixbuf = None

        # remove the audio pipeline
        self._audio.get_bus().remove_signal_watch()
        self._audio.set_state(Gst.State.NULL)  # synchronous
        self._audio = None
        return False

    def record_video(self, quality):
        logger.debug('record_video')
        if not self._cameras:
            return

        # take a photograph
        self._video_pixbuf = self._pixbuf.copy()

        # stop the live view
        self._pipeline.set_state(Gst.State.NULL)  # synchronous
        self._pixbuf = None

        # make a pipeline to record video and audio to file
        args = {}
        args['ogv'] = os.path.join(Instance.instancePath, "output.ogv")
        args['src'] = 'v4l2src device={0}'.format(self._camera)
        if self._camera == 'test':
            args['src'] = 'videotestsrc'

        cmd = '{src} name=vsrc ' \
            '! video/x-raw,width=640,height=480 ' \
            '! tee name=tee ' \
            'tee.! videorate max-rate=2 ! videoconvert ' \
            '! queue leaky=2 ' \
            '! autovideosink sync=false ' \
            'tee.! videorate max-rate=10 ! videoconvert ' \
            '! queue max-size-time=30000000000 ' \
            'max-size-bytes=0 max-size-buffers=0 ' \
            '! theoraenc name=theora ' \
            '! mux. ' \
            'autoaudiosrc name=asrc ' \
            '! audiorate ' \
            '! audioconvert ' \
            '! queue max-size-time=30000000000 ' \
            'max-size-bytes=0 max-size-buffers=0 ' \
            '! vorbisenc name=vorbis ! mux. ' \
            'oggmux name=mux ! filesink location={ogv}'
        self._video = Gst.parse_launch(cmd.format(**args))

        # attach useful tags
        taglist = self._get_tags(constants.TYPE_VIDEO)
        if self._video_pixbuf:
            pixbuf_b64 = utils.getStringEncodedFromPixbuf(self._video_pixbuf)
            taglist.add_value(Gst.TagMergeMode.APPEND,
                              Gst.TAG_EXTENDED_COMMENT,
                              "coverart=" + str(pixbuf_b64))

        theora = self._video.get_by_name('theora')
        vorbis = self._video.get_by_name('vorbis')
        vorbis.merge_tags(taglist, Gst.TagMergeMode.REPLACE_ALL)

        # set quality
        if quality == 0:
            theora.props.quality = 24
            vorbis.props.quality = 0.2
        if quality == 1:
            theora.props.quality = 52
            vorbis.props.quality = 0.4

        # catch end of stream
        bus = self._video.get_bus()
        bus.add_signal_watch()

        def on_message_cb(bus, msg, ogv):
            if msg.type == Gst.MessageType.EOS:
                logger.debug('record_video.on_message_cb Gst.MessageType.EOS')
                GLib.idle_add(self._stop_recording_video, ogv)
                return

            return self._on_message_cb(bus, msg)

        bus.connect('message', on_message_cb, args['ogv'])

        self._catch_window(bus, self.activity.set_glive_sink)

        # start video pipeline recording
        self._video.set_state(Gst.State.PLAYING)  # asynchronous

    def stop_recording_video(self):
        logger.debug('stop_recording_video')
        if not self._cameras:
            return

        # ask for stream to end
        self._video.get_by_name('vsrc').send_event(Gst.Event.new_eos())
        self._video.get_by_name('asrc').send_event(Gst.Event.new_eos())

    def _stop_recording_video(self, ogv):
        logger.debug('_stop_recording_video')

        # save video file
        self.model.save_video(ogv, self._video_pixbuf)
        self._video_pixbuf = None

        # remove the video pipeline
        self._video.get_bus().remove_signal_watch()
        self._video.set_state(Gst.State.NULL)  # synchronous
        self._video = None
        return False

    def _get_tags(self, mediatype):
        tl = Gst.TagList.new_empty()

        def _set(tag, value):
            tl.add_value(Gst.TagMergeMode.APPEND, tag, value)
        _set(Gst.TAG_ARTIST, self.model.get_nickname())
        _set(Gst.TAG_COMMENT, "olpc")
        _set(Gst.TAG_ALBUM, "olpc")
        date = GLib.Date.new()
        date.set_time_t(time.time())
        _set(Gst.TAG_DATE, date)
        stringType = constants.MEDIA_INFO[mediatype]['istr']

        # Translators: photo by photographer, e.g. "Photo by Mary"
        title = _('%(type)s by %(name)s') % {'type': stringType,
                                             'name': self.model.get_nickname()}
        _set(Gst.TAG_TITLE, title)
        return tl
